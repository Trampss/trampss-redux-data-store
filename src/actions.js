const scope = action => `@trampss/${action.toUpperCase()}`

export const SET = prefix => scope(`SET_${prefix}`)
export const set = prefix => payload => ({ type: SET(prefix), payload })

export const RESET = prefix => scope(`RESET_${prefix}`)
export const reset = prefix => () => ({ type: RESET(prefix) })

export const ADD = prefix => scope(`ADD_${prefix}`)
export const add = prefix => payload => ({ type: ADD(prefix), payload })

export const UPDATE = prefix => scope(`UPDATE_${prefix}`)
export const update = prefix => payload => ({ type: UPDATE(prefix), payload })

export const REMOVE = prefix => scope(`REMOVE_${prefix}`)
export const remove = prefix => key => ({ type: REMOVE(prefix), payload: key })

export const ADD_OR_UPDATE = prefix => scope(`ADD_OR_UPDATE_${prefix}`)
export const addOrUpdate = prefix => payload => ({ type: ADD_OR_UPDATE(prefix), payload })
