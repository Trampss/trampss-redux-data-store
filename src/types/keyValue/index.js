import middleware from './keyValue.middleware'

export default {
  middlewares: [middleware],
  actions: ['set', 'add', 'reset', 'remove', 'update'],
  selectorsEnabled: ['getState', 'getKeys', 'getAsArray', 'getLength', 'isInitialized', 'get', 'getBy'],
}
