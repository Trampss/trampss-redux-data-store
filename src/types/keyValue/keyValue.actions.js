import {
  set, SET,
  add, ADD,
  reset, RESET,
  remove, REMOVE,
  update, UPDATE,
  addOrUpdate, ADD_OR_UPDATE,
} from '../../actions'

export {
  set, SET,
  add, ADD,
  reset, RESET,
  remove, REMOVE,
  update, UPDATE,
  addOrUpdate, ADD_OR_UPDATE,
}
