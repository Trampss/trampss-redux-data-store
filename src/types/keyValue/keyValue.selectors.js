const getFromPath = (data, path) => path.split('.').reduce(
  (curr, sub) => curr && curr[sub],
  data,
)

const memoize = (callback) => {
  const memory = {
    data: undefined,
    result: undefined,
  }

  return (data) => {
    if (memory.data !== data) {
      memory.data = data
      memory.result = callback(data)
    }

    return memory.result
  }
}

export default (options) => {
  // get reducer state from rootState
  const getState = (rootState) => {
    const { path, name } = options
    let subState = rootState

    if (path !== undefined && path.length > 0) subState = getFromPath(rootState, path)

    return subState[name]
  }

  // basic selectors
  const getData = rootState => getState(rootState).data
  const getMap = memoize(data => new Map(data))
  const mapToKeys = memoize(data => Array.from(getMap(data).keys()))
  const mapToValues = memoize(data => Array.from(getMap(data).values()))
  const getKeys = rootState => mapToKeys(getData(rootState))
  const getObject = memoize((data) => {
    const object = {}
    data.forEach(([key, value]) => {
      object[key] = value
    })
    return object
  })
  const get = keys => (rootState) => {
    const data = getData(rootState)
    // all data
    if (keys === undefined || keys === null) return getObject(data)
    // by key(s)
    const map = getMap(data)
    if (Array.isArray(keys)) return keys.map(key => map.get(key))
    return map.get(keys)
  }
  const getBy = (path, filters) => (rootState) => {
    const values = mapToValues(getMap(getData(rootState)))

    if (Array.isArray(filters)) return values.filter(d => filters.includes(getFromPath(d, path)))
    return values.filter(d => filters === getFromPath(d, path))
  }

  return {
    getState,
    get,
    getKeys,
    getBy,
    getAsArray: rootState => mapToValues(getData(rootState)),
    getLength: rootState => getData(rootState).length,
    hasKey: key => rootState => getMap(getData(rootState)).has(key),
    isInitialized: rootState => getState(rootState).initialized,
  }
}
