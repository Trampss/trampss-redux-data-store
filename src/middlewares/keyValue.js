import { keyBy, without, uniq, omit } from 'lodash'
import { SET, ADD, UPDATE, REMOVE, RESET } from '../actions'

export const initState = { data: {}, keys: [], array: [], initialized: false }

const keyAlreadyExists =
  state => (key, instanceKey) => state.array.find(o => o[key] === instanceKey)

const reducer = key => prefix => (/* defaultData */) =>
  (state = initState, { type, payload } = {}) => {
    switch (type) {
      case SET(prefix):
        return {
          data: keyBy(payload, key),
          keys: uniq(payload.map(element => element[key])),
          array: payload,
          initialized: true,
        }
      case ADD(prefix): {
        let array
        const instanceKey = payload[key]
        if (!keyAlreadyExists(state)(key, instanceKey)) {
          array = [...state.array, payload]
        } else {
          array = state.array.map(o => (o[key] === instanceKey ? payload : o))
        }

        return {
          ...state,
          data: { ...state.data, [instanceKey]: payload },
          keys: uniq([...state.keys, instanceKey]),
          array,
          initialized: true,
        }
      }
      case UPDATE(prefix): {
        const instanceKey = payload[key]
        if (!keyAlreadyExists(state)(key, instanceKey)) return state
        return {
          ...state,
          data: { ...state.data, [instanceKey]: { ...state.data[instanceKey], ...payload } },
          array: state.array.map(o => (o[key] === instanceKey ? { ...o, ...payload } : o)),
        }
      }
      case REMOVE(prefix):
        return {
          ...state,
          data: omit(state.data, [payload]),
          keys: without(state.keys, payload),
          array: state.array ? state.array.filter(o => o[key] !== payload) : [],
        }
      case RESET(prefix):
        return initState
      default:
        return state
    }
  }

export default key => prefix => defaultData => (ctx = {}) => ({
  ...ctx,
  state: reducer(key)(prefix)(defaultData)(ctx.state, ctx.action),
})
