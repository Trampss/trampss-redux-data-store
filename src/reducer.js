import { keyBy, without, uniq, omit } from 'lodash'
import { SET, ADD, DEL, RESET } from './actions'

export const initState = { data: {}, keys: [], array: [], nb: 0, initialized: false }

export default key => prefix =>
  (state = initState, { type = 'UNKONWN', payload } = {}) => {
    switch (type) {
      case SET(prefix):
        return {
          data: keyBy(payload, key),
          keys: payload.map(element => element[key]),
          array: payload,
          nb: payload.length,
          initialized: true,
        }
      case ADD(prefix):
        return {
          ...state,
          data: { ...state.data, [payload[key]]: payload },
          keys: uniq([...state.keys, payload[key]]),
          array: [...state.array, payload],
          nb: state.keys.length + 1,
          initialized: true,
        }
      case DEL(prefix):
        return {
          ...state,
          data: omit(state.data, [payload]),
          keys: without(state.keys, payload),
          array: state.array ? state.array.filter(o => o[key] !== payload) : [],
          nb: state.keys.length - 1,
        }
      case RESET(prefix):
        return initState
      default:
        return state
    }
  }
