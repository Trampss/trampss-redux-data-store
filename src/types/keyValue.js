import keyValue from '../middlewares/keyValue'

export default {
  middlewares: [keyValue],
  actions: ['set', 'add', 'reset', 'remove', 'update'],
  selectors: ['getState', 'getKeys', 'getAsArray', 'getLength', 'isInitialized', 'get', 'getBy'],
}
