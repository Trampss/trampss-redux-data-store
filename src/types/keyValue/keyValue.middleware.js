import { keyBy, without, uniq, omit, orderBy, get, isObjectLike, isString, flatten } from 'lodash'
import { SET, ADD, UPDATE, REMOVE, RESET, ADD_OR_UPDATE, REPLACE, ORDER_BY } from './keyValue.actions'

export const initState = {
  data: {},
  keys: [],
  array: [],
  initialized: false,
}

const keyAlreadyExists =
  state => (key, instanceKey) => state.array.find(o => o[key] === instanceKey)

const set = (key, payload) => ({
  data: keyBy(payload, key),
  keys: uniq(payload.map(element => element[key])),
  array: payload,
  initialized: true,
})

const add = (key, state, payload) => {
  let array
  const instanceKey = payload[key]
  if (!keyAlreadyExists(state)(key, instanceKey)) {
    array = [...state.array, payload]
  } else {
    array = state.array.map(o => (o[key] === instanceKey ? payload : o))
  }

  return {
    ...state,
    data: { ...state.data, [instanceKey]: payload },
    keys: uniq([...state.keys, instanceKey]),
    array,
    initialized: true,
  }
}

const update = (key, state, payload) => {
  const instanceKey = payload[key]

  return {
    ...state,
    data: { ...state.data, [instanceKey]: { ...state.data[instanceKey], ...payload } },
    array: state.array.map(o => (o[key] === instanceKey ? { ...o, ...payload } : o)),
  }
}

const replace = (key, state, payload) => {
  const instanceKey = payload[key]

  return {
    ...state,
    data: { ...state.data, [instanceKey]: payload },
    array: state.array.map(o => (o[key] === instanceKey ? payload : o)),
  }
}

const defaultState =
  (key, defaultData) => (defaultData !== undefined ? set(key, defaultData) : initState)

const reducer = key => prefix => name => defaultData =>
  (state = defaultState(key, defaultData), { type, payload } = {}) => {
    switch (type) {
      case SET(prefix)(name): return set(key, payload)
      case ADD(prefix)(name): {
      if (Array.isArray(payload))
        return payload.reduce((newState, entity) => add(key, newState, entity), state)
      return add(key, state, payload)
      }
      case ADD_OR_UPDATE(prefix)(name): {
      if (Array.isArray(payload))
        return payload.reduce(
          (newState, entity) =>
            keyAlreadyExists(state)(key, entty[key])
              ? update(key, state, entity)
              : add(key, newState, entity),
          state,
        )
        if (!keyAlreadyExists(state)(key, payload[key])) return add(key, state, payload)
        return update(key, state, payload)
      }
      case UPDATE(prefix)(name): {
      if (Array.isArray(payload))
        return payload.reduce(
          (newState, entity) =>
            keyAlreadyExists(state)(key, entty[key]) ? update(key, state, entity) : newState,
          state,
        )
        if (!keyAlreadyExists(state)(key, payload[key])) return state
        return update(key, state, payload)
      }
      case REPLACE(prefix)(name): {
      if (Array.isArray(payload))
        return payload.reduce(
          (newState, entity) =>
            keyAlreadyExists(state)(key, entty[key]) ? replace(key, state, entity) : newState,
          state,
        )
        if (!keyAlreadyExists(state)(key, payload[key])) return state
        return replace(key, state, payload)
      }
      case ORDER_BY(prefix)(name): {
        let by = payload
        let orders = 'asc'
        if (isObjectLike(payload)) {
          by = payload.by // eslint-disable-line prefer-destructuring
          orders = payload.desc ? 'desc' : 'asc'
        }
        const arraySorted = orderBy(
          state.array,
          isString(by) ? p => get(p, by) : by,
          orders,
        )
        return {
          ...state,
          array: arraySorted,
          keys: uniq(arraySorted.map(element => element[key])),
        }
      }
      case REMOVE(prefix)(name): {
        const removeIds = flatten([payload])
        return {
          ...state,
          data: omit(state.data, removeIds),
          keys: without(state.keys, ...removeIds),
          array: state.array ? state.array.filter(o => !removeIds.includes(o[key])) : [],
        }
      }
      case RESET(prefix)(name):
        return defaultState(key, defaultData)
      default:
        return state
    }
  }

export default key => prefix => name => defaultData => (ctx = {}) => ({
  ...ctx,
  state: reducer(key)(prefix)(name)(defaultData)(ctx.state, ctx.action),
})
