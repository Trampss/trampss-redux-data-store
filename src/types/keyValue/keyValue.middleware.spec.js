/* eslint-env jest */
import { set, add, update, remove, reset, addOrUpdate, orderBy } from '../../actions'
import { initState } from '../../reducer'
import keyValue from './keyValue.middleware'

const prefix = 'testPrefix'
const Element = code => ({ code, some: 'other', infos: code, subinfos: { info: code } })
const state = {
  data: {
    elm2: Element('elm2'),
    elm1: Element('elm1'),
    elm3: Element('elm3'),
  },
  keys: ['elm2', 'elm1', 'elm3'],
  array: [Element('elm2'), Element('elm1'), Element('elm3')],
  initialized: false,
}

describe('middlewares/keyValue', () => {
  // Use the factory to create a new reducer named 'testPrefix'
  // It uses 'code' as key in givent elements
  const testPrefix = keyValue('code')(prefix)(/* defaultData */)

  it('should initialize', () => expect(testPrefix()).toMatchSnapshot())

  it('should add element [elm4] on initial store', () => expect(
    testPrefix({
      state: initState,
      action: add(prefix)(Element('elm4')),
    }),
  ).toMatchSnapshot())

  it('should add element [elm4]', () => expect(
    testPrefix({
      state,
      action: add(prefix)(Element('elm4')),
    }),
  ).toMatchSnapshot())

  it('should set elements [elm1,elm2]', () => expect(
    testPrefix({
      state,
      action: set(prefix)([Element('elm1'), Element('elm2')]),
    }),
  ).toMatchSnapshot())

  it('should remove element [elm1]', () => expect(
    testPrefix({
      state,
      action: remove(prefix)('elm1'),
    }),
  ).toMatchSnapshot())

  it('should remove many element [elm1, elm2]', () => expect(
    testPrefix({
      state,
      action: remove(prefix)(['elm1', 'elm2']),
    }),
  ).toMatchSnapshot())

  it('should reset state', () => expect(
    testPrefix({
      state,
      action: reset(prefix)(),
    }),
  ).toMatchSnapshot())

  it('should not duplicate object with same key [elm2]', () => expect(
    testPrefix({
      state,
      action: add(prefix)(Element('elm2')),
    }),
  ).toMatchSnapshot())

  it('should update an element [elm3]', () => expect(
    testPrefix({
      state,
      action: update(prefix)({ ...Element('elm3'), some: 'other UPDATE' }),
    }),
  ).toMatchSnapshot())

  it('should not update element if not found [elm12]', () => expect(
    testPrefix({
      state,
      action: update(prefix)(Element('elm12')),
    }),
  ).toMatchSnapshot())

  it('should update element [elm3]', () => expect(
    testPrefix({
      state,
      action: addOrUpdate(prefix)({ ...Element('elm3'), some: 'other ADD OR UPDATE' }),
    }),
  ).toMatchSnapshot())

  it('should add element if not found [elm12]', () => expect(
    testPrefix({
      state,
      action: addOrUpdate(prefix)(Element('elm12')),
    }),
  ).toMatchSnapshot())

  it('should order elements by identity -code asc-', () => expect(
    testPrefix({
      state,
      action: orderBy(prefix)('code'),
    }),
  ).toMatchSnapshot())

  it('should order elements by identity -sub info asc-', () => expect(
    testPrefix({
      state,
      action: orderBy(prefix)('subinfos.info'),
    }),
  ).toMatchSnapshot())

  it('should order elements by identity -code desc-', () => expect(
    testPrefix({
      state,
      action: orderBy(prefix)({ by: 'code', desc: true }),
    }),
  ).toMatchSnapshot())

  it('should order elements by identity -sub info desc-', () => expect(
    testPrefix({
      state,
      action: orderBy(prefix)({ by: 'subinfos.info', desc: true }),
    }),
  ).toMatchSnapshot())

  it('should order elements with function -code desc-', () => expect(
    testPrefix({
      state,
      action: orderBy(prefix)({ by: e => e.code, desc: true }),
    }),
  ).toMatchSnapshot())

  it('should order elements with function -code asc (objec mode)-', () => expect(
    testPrefix({
      state,
      action: orderBy(prefix)({ by: e => e.code, desc: false }),
    }),
  ).toMatchSnapshot())

  it('should order elements with function -code asc-', () => expect(
    testPrefix({
      state,
      action: orderBy(prefix)(e => e.code),
    }),
  ).toMatchSnapshot())
})
