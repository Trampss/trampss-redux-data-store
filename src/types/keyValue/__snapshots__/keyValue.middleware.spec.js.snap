// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`middlewares/keyValue should add element [elm4] 1`] = `
Object {
  "action": Object {
    "payload": Object {
      "code": "elm4",
      "infos": "elm4",
      "some": "other",
    },
    "type": "@trampss/ADD_TESTPREFIX",
  },
  "state": Object {
    "array": Array [
      Object {
        "code": "elm2",
        "infos": "elm2",
        "some": "other",
      },
      Object {
        "code": "elm1",
        "infos": "elm1",
        "some": "other",
      },
      Object {
        "code": "elm3",
        "infos": "elm3",
        "some": "other",
      },
      Object {
        "code": "elm4",
        "infos": "elm4",
        "some": "other",
      },
    ],
    "data": Object {
      "elm1": Object {
        "code": "elm1",
        "infos": "elm1",
        "some": "other",
      },
      "elm2": Object {
        "code": "elm2",
        "infos": "elm2",
        "some": "other",
      },
      "elm3": Object {
        "code": "elm3",
        "infos": "elm3",
        "some": "other",
      },
      "elm4": Object {
        "code": "elm4",
        "infos": "elm4",
        "some": "other",
      },
    },
    "initialized": true,
    "keys": Array [
      "elm2",
      "elm1",
      "elm3",
      "elm4",
    ],
  },
}
`;

exports[`middlewares/keyValue should add element [elm4] on initial store 1`] = `
Object {
  "action": Object {
    "payload": Object {
      "code": "elm4",
      "infos": "elm4",
      "some": "other",
    },
    "type": "@trampss/ADD_TESTPREFIX",
  },
  "state": Object {
    "array": Array [
      Object {
        "code": "elm4",
        "infos": "elm4",
        "some": "other",
      },
    ],
    "data": Object {
      "elm4": Object {
        "code": "elm4",
        "infos": "elm4",
        "some": "other",
      },
    },
    "initialized": true,
    "keys": Array [
      "elm4",
    ],
  },
}
`;

exports[`middlewares/keyValue should add element if not found [elm12] 1`] = `
Object {
  "action": Object {
    "payload": Object {
      "code": "elm12",
      "infos": "elm12",
      "some": "other",
    },
    "type": "@trampss/ADD_OR_UPDATE_TESTPREFIX",
  },
  "state": Object {
    "array": Array [
      Object {
        "code": "elm2",
        "infos": "elm2",
        "some": "other",
      },
      Object {
        "code": "elm1",
        "infos": "elm1",
        "some": "other",
      },
      Object {
        "code": "elm3",
        "infos": "elm3",
        "some": "other",
      },
      Object {
        "code": "elm12",
        "infos": "elm12",
        "some": "other",
      },
    ],
    "data": Object {
      "elm1": Object {
        "code": "elm1",
        "infos": "elm1",
        "some": "other",
      },
      "elm12": Object {
        "code": "elm12",
        "infos": "elm12",
        "some": "other",
      },
      "elm2": Object {
        "code": "elm2",
        "infos": "elm2",
        "some": "other",
      },
      "elm3": Object {
        "code": "elm3",
        "infos": "elm3",
        "some": "other",
      },
    },
    "initialized": true,
    "keys": Array [
      "elm2",
      "elm1",
      "elm3",
      "elm12",
    ],
  },
}
`;

exports[`middlewares/keyValue should initialize 1`] = `
Object {
  "state": Object {
    "array": Array [],
    "data": Object {},
    "initialized": false,
    "keys": Array [],
  },
}
`;

exports[`middlewares/keyValue should not duplicate object with same key [elm2] 1`] = `
Object {
  "action": Object {
    "payload": Object {
      "code": "elm2",
      "infos": "elm2",
      "some": "other",
    },
    "type": "@trampss/ADD_TESTPREFIX",
  },
  "state": Object {
    "array": Array [
      Object {
        "code": "elm2",
        "infos": "elm2",
        "some": "other",
      },
      Object {
        "code": "elm1",
        "infos": "elm1",
        "some": "other",
      },
      Object {
        "code": "elm3",
        "infos": "elm3",
        "some": "other",
      },
    ],
    "data": Object {
      "elm1": Object {
        "code": "elm1",
        "infos": "elm1",
        "some": "other",
      },
      "elm2": Object {
        "code": "elm2",
        "infos": "elm2",
        "some": "other",
      },
      "elm3": Object {
        "code": "elm3",
        "infos": "elm3",
        "some": "other",
      },
    },
    "initialized": true,
    "keys": Array [
      "elm2",
      "elm1",
      "elm3",
    ],
  },
}
`;

exports[`middlewares/keyValue should not update element if not found [elm12] 1`] = `
Object {
  "action": Object {
    "payload": Object {
      "code": "elm12",
      "infos": "elm12",
      "some": "other",
    },
    "type": "@trampss/UPDATE_TESTPREFIX",
  },
  "state": Object {
    "array": Array [
      Object {
        "code": "elm2",
        "infos": "elm2",
        "some": "other",
      },
      Object {
        "code": "elm1",
        "infos": "elm1",
        "some": "other",
      },
      Object {
        "code": "elm3",
        "infos": "elm3",
        "some": "other",
      },
    ],
    "data": Object {
      "elm1": Object {
        "code": "elm1",
        "infos": "elm1",
        "some": "other",
      },
      "elm2": Object {
        "code": "elm2",
        "infos": "elm2",
        "some": "other",
      },
      "elm3": Object {
        "code": "elm3",
        "infos": "elm3",
        "some": "other",
      },
    },
    "initialized": false,
    "keys": Array [
      "elm2",
      "elm1",
      "elm3",
    ],
  },
}
`;

exports[`middlewares/keyValue should order elements by identity -code asc- 1`] = `
Object {
  "action": Object {
    "payload": "code",
    "type": "@trampss/ORDER_BY_TESTPREFIX",
  },
  "state": Object {
    "array": Array [
      Object {
        "code": "elm1",
        "infos": "elm1",
        "some": "other",
      },
      Object {
        "code": "elm2",
        "infos": "elm2",
        "some": "other",
      },
      Object {
        "code": "elm3",
        "infos": "elm3",
        "some": "other",
      },
    ],
    "data": Object {
      "elm1": Object {
        "code": "elm1",
        "infos": "elm1",
        "some": "other",
      },
      "elm2": Object {
        "code": "elm2",
        "infos": "elm2",
        "some": "other",
      },
      "elm3": Object {
        "code": "elm3",
        "infos": "elm3",
        "some": "other",
      },
    },
    "initialized": false,
    "keys": Array [
      "elm1",
      "elm2",
      "elm3",
    ],
  },
}
`;

exports[`middlewares/keyValue should order elements by identity -code desc- 1`] = `
Object {
  "action": Object {
    "payload": Object {
      "by": "code",
      "desc": true,
    },
    "type": "@trampss/ORDER_BY_TESTPREFIX",
  },
  "state": Object {
    "array": Array [
      Object {
        "code": "elm3",
        "infos": "elm3",
        "some": "other",
      },
      Object {
        "code": "elm2",
        "infos": "elm2",
        "some": "other",
      },
      Object {
        "code": "elm1",
        "infos": "elm1",
        "some": "other",
      },
    ],
    "data": Object {
      "elm1": Object {
        "code": "elm1",
        "infos": "elm1",
        "some": "other",
      },
      "elm2": Object {
        "code": "elm2",
        "infos": "elm2",
        "some": "other",
      },
      "elm3": Object {
        "code": "elm3",
        "infos": "elm3",
        "some": "other",
      },
    },
    "initialized": false,
    "keys": Array [
      "elm3",
      "elm2",
      "elm1",
    ],
  },
}
`;

exports[`middlewares/keyValue should order elements with function -code asc- 1`] = `
Object {
  "action": Object {
    "payload": [Function],
    "type": "@trampss/ORDER_BY_TESTPREFIX",
  },
  "state": Object {
    "array": Array [
      Object {
        "code": "elm1",
        "infos": "elm1",
        "some": "other",
      },
      Object {
        "code": "elm2",
        "infos": "elm2",
        "some": "other",
      },
      Object {
        "code": "elm3",
        "infos": "elm3",
        "some": "other",
      },
    ],
    "data": Object {
      "elm1": Object {
        "code": "elm1",
        "infos": "elm1",
        "some": "other",
      },
      "elm2": Object {
        "code": "elm2",
        "infos": "elm2",
        "some": "other",
      },
      "elm3": Object {
        "code": "elm3",
        "infos": "elm3",
        "some": "other",
      },
    },
    "initialized": false,
    "keys": Array [
      "elm1",
      "elm2",
      "elm3",
    ],
  },
}
`;

exports[`middlewares/keyValue should order elements with function -code desc- 1`] = `
Object {
  "action": Object {
    "payload": Object {
      "by": [Function],
      "desc": true,
    },
    "type": "@trampss/ORDER_BY_TESTPREFIX",
  },
  "state": Object {
    "array": Array [
      Object {
        "code": "elm3",
        "infos": "elm3",
        "some": "other",
      },
      Object {
        "code": "elm2",
        "infos": "elm2",
        "some": "other",
      },
      Object {
        "code": "elm1",
        "infos": "elm1",
        "some": "other",
      },
    ],
    "data": Object {
      "elm1": Object {
        "code": "elm1",
        "infos": "elm1",
        "some": "other",
      },
      "elm2": Object {
        "code": "elm2",
        "infos": "elm2",
        "some": "other",
      },
      "elm3": Object {
        "code": "elm3",
        "infos": "elm3",
        "some": "other",
      },
    },
    "initialized": false,
    "keys": Array [
      "elm3",
      "elm2",
      "elm1",
    ],
  },
}
`;

exports[`middlewares/keyValue should remove element [elm1] 1`] = `
Object {
  "action": Object {
    "payload": "elm1",
    "type": "@trampss/REMOVE_TESTPREFIX",
  },
  "state": Object {
    "array": Array [
      Object {
        "code": "elm2",
        "infos": "elm2",
        "some": "other",
      },
      Object {
        "code": "elm3",
        "infos": "elm3",
        "some": "other",
      },
    ],
    "data": Object {
      "elm2": Object {
        "code": "elm2",
        "infos": "elm2",
        "some": "other",
      },
      "elm3": Object {
        "code": "elm3",
        "infos": "elm3",
        "some": "other",
      },
    },
    "initialized": false,
    "keys": Array [
      "elm2",
      "elm3",
    ],
  },
}
`;

exports[`middlewares/keyValue should reset state 1`] = `
Object {
  "action": Object {
    "type": "@trampss/RESET_TESTPREFIX",
  },
  "state": Object {
    "array": Array [],
    "data": Object {},
    "initialized": false,
    "keys": Array [],
  },
}
`;

exports[`middlewares/keyValue should set elements [elm1,elm2] 1`] = `
Object {
  "action": Object {
    "payload": Array [
      Object {
        "code": "elm1",
        "infos": "elm1",
        "some": "other",
      },
      Object {
        "code": "elm2",
        "infos": "elm2",
        "some": "other",
      },
    ],
    "type": "@trampss/SET_TESTPREFIX",
  },
  "state": Object {
    "array": Array [
      Object {
        "code": "elm1",
        "infos": "elm1",
        "some": "other",
      },
      Object {
        "code": "elm2",
        "infos": "elm2",
        "some": "other",
      },
    ],
    "data": Object {
      "elm1": Object {
        "code": "elm1",
        "infos": "elm1",
        "some": "other",
      },
      "elm2": Object {
        "code": "elm2",
        "infos": "elm2",
        "some": "other",
      },
    },
    "initialized": true,
    "keys": Array [
      "elm1",
      "elm2",
    ],
  },
}
`;

exports[`middlewares/keyValue should update an element [elm3] 1`] = `
Object {
  "action": Object {
    "payload": Object {
      "code": "elm3",
      "infos": "elm3",
      "some": "other UPDATE",
    },
    "type": "@trampss/UPDATE_TESTPREFIX",
  },
  "state": Object {
    "array": Array [
      Object {
        "code": "elm2",
        "infos": "elm2",
        "some": "other",
      },
      Object {
        "code": "elm1",
        "infos": "elm1",
        "some": "other",
      },
      Object {
        "code": "elm3",
        "infos": "elm3",
        "some": "other UPDATE",
      },
    ],
    "data": Object {
      "elm1": Object {
        "code": "elm1",
        "infos": "elm1",
        "some": "other",
      },
      "elm2": Object {
        "code": "elm2",
        "infos": "elm2",
        "some": "other",
      },
      "elm3": Object {
        "code": "elm3",
        "infos": "elm3",
        "some": "other UPDATE",
      },
    },
    "initialized": false,
    "keys": Array [
      "elm2",
      "elm1",
      "elm3",
    ],
  },
}
`;

exports[`middlewares/keyValue should update element [elm3] 1`] = `
Object {
  "action": Object {
    "payload": Object {
      "code": "elm3",
      "infos": "elm3",
      "some": "other ADD OR UPDATE",
    },
    "type": "@trampss/ADD_OR_UPDATE_TESTPREFIX",
  },
  "state": Object {
    "array": Array [
      Object {
        "code": "elm2",
        "infos": "elm2",
        "some": "other",
      },
      Object {
        "code": "elm1",
        "infos": "elm1",
        "some": "other",
      },
      Object {
        "code": "elm3",
        "infos": "elm3",
        "some": "other ADD OR UPDATE",
      },
    ],
    "data": Object {
      "elm1": Object {
        "code": "elm1",
        "infos": "elm1",
        "some": "other",
      },
      "elm2": Object {
        "code": "elm2",
        "infos": "elm2",
        "some": "other",
      },
      "elm3": Object {
        "code": "elm3",
        "infos": "elm3",
        "some": "other ADD OR UPDATE",
      },
    },
    "initialized": false,
    "keys": Array [
      "elm2",
      "elm1",
      "elm3",
    ],
  },
}
`;
